### PR TITLE
Brave Talk - share a tab (during a call)

### PR DESCRIPTION
**Pull request by fallaciousreasoning**
Resolves [#22349](https://github.com/brave/brave-browser/issues/22349)

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/views/toolbar/brave_toolbar_view.cc [hunk 1]
```diff
@@ -11,9 +11,12 @@
 
 #include "base/bind.h"
 #include "brave/app/brave_command_ids.h"
+#include "brave/browser/brave_wallet/brave_wallet_context_utils.h"
```
Comment (fallaciousreasoning) [inline browser/ui/views/toolbar/brave_toolbar_view.cc:14]: Note: I haven't reverted the change to this file as we might be adding the share button here.

File: components/brave_talk/browser/brave_talk_frame_host.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+#include "content/browser/bad_message.h"
+#include "content/browser/renderer_host/frame_tree_node.h"
+#include "content/browser/renderer_host/navigation_controller_impl.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents.h"
+#include "third_party/blink/public/common/tokens/tokens.h"
+
+namespace brave_talk {
+class FrameFinder {
```
Comment (fallaciousreasoning) [inline components/brave_talk/browser/brave_talk_frame_host.cc:33]: This `FrameFinder` stuff can all be avoided if I use `RenderFrameHostImpl::FindAndVerifyChild`, which takes a `blink::FrameToken` directly. However, I vaguely remember that depending on stuff in `content/browser` is a bit of a nono. 

I'd be interested to hear someone who knows more's thoughts (like, maybe this is a bit of a smell).

File: components/brave_talk/renderer/brave_talk_frame_js_handler.cc [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  if (!global->Get(context, gin::StringToV8(isolate, "brave"))
+           .ToLocal(&brave_value) ||
+      !brave_value->IsObject()) {
+    brave_obj = v8::Object::New(isolate);
+    global->Set(context, gin::StringToSymbol(isolate, "brave"), brave_obj)
+        .Check();
+  } else {
+    brave_obj = brave_value->ToObject(context).ToLocalChecked();
+  }
+  BindFunctionToObject(
```
Comment (fallaciousreasoning) [inline components/brave_talk/renderer/brave_talk_frame_js_handler.cc:74]: Note: Currently this forces you to specify a second argument, but I imagine we can just override this to make it optional, if we decide we like this API shape.

File: browser/brave_talk/brave_talk_service.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#include "content/public/browser/render_frame_host.h"
+#include "content/public/browser/web_contents_observer.h"
+
+namespace content {
+class WebContents;
+}
+
+namespace brave_talk {
+
+// Service for managing requests to |brave.beginAdvertiseShareDisplayMedia|. Note:
```
Comment (fallaciousreasoning) [inline browser/brave_talk/brave_talk_service.h:22]: @petemill what do you think of this approach/tradeoff? It's definitely the simplest, as it means we don't need to implement any new UI for picking which frame to share with, and least fraught. However, we definitely wouldn't be able to do this if we were planning on desigining an API for the wider web (any website could steal the tab share button, any time).

**Comment by fallaciousreasoning** [comment]
@petemill would you mind taking a glance at this to see if it's on the right track?

File: app/brave_generated_resources.grd [hunk 5]
```diff
@@ -429,6 +429,17 @@ Or change later at <ph name="SETTINGS_EXTENIONS_LINK">$2<ex>brave://settings/ext
...
 
+      <!-- Brave Talk Tab Sharing -->
+      <message name="IDS_BRAVE_TALK_SHARE_TAB_BUTTON_TOOLTIP" desc="The description of the share tab entry in the sharing hub">
+        Share Tab
+      </message>
+      <message name="IDS_BRAVE_TALK_SHARE_TAB_CONFIRM_PROMPT_TITLE" desc="The title of the confirm popup which shows when the user clicks the share tab button">
+        Share this tab?
+      </message>
+      <message name="IDS_BRAVE_TALK_SHARE_TAB_CONFIRM_PROMPT_MESSAGE" desc="The message in the confirm popup which shows when the user clicks the share tab button">
+        The contents of this tab will be shared with your active Brave Talk call
```
Comment (diracdeltas) [inline app/brave_generated_resources.grd:440]: what happens if a user is in two calls at once?

**Comment by petemill** [comment]
- Have the infobar show up
image

- Move icon to the location bar (to the left of the brave actions container)

File: components/brave_talk/browser/brave_talk_frame_host.cc [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+#include "content/public/browser/web_contents.h"
+#include "third_party/blink/public/common/tokens/tokens.h"
+
+namespace brave_talk {
+
+namespace {
+content::RenderFrameHost* FindFrameForToken(
+    const absl::optional<::blink::FrameToken>& frame_token,
+    content::WebContents* contents) {
+  auto* main_frame =
```
Comment (fallaciousreasoning) [inline components/brave_talk/browser/brave_talk_frame_host.cc:38]: @simonhong would you mind taking a look at in particular. I feel like this is probably a layering violation, but I also couldn't find an API for getting a RenderFrameHost from a FrameToken (without knowing the processID). 

Happy to change the way this works if you have a suggestion.

Comment (simonhong) [reply components/brave_talk/browser/brave_talk_frame_host.cc:38]: Well, I think this method doesn't have any layer violation.
However, using BraveTalkService here is layer violation.

**Pull request closed without merging**